### PR TITLE
gh-114271: Fix race in `Thread.join()`

### LINKED ISSUE #116372: Race in `Thread.join()`
# Bug report

### Bug description:

There is a race between when `Thread._tstate_lock` is released[^1] in `Thread._wait_for_tstate_lock()` and when `Thread._stop()` asserts[^2] that it is unlocked. Consider the following execution involving threads A, B, and C:

1. A starts.
2. B joins A, blocking on its `_tstate_lock`.
3. C joins A, blocking on its `_tstate_lock`.
4. A finishes and releases its `_tstate_lock`.
5. B acquires A's `_tstate_lock` in `_wait_for_tstate_lock()`, releases it, but is swapped out before calling `_stop()`.
6. C is scheduled, acquires A's `_tstate_lock` in `_wait_for_tstate_lock()` but is swapped out before releasing it.
7. B is scheduled, calls `_stop()`, which asserts that A's `_tstate_lock` is not held. However, C holds it, so the assertion fails.

The race can be reproduced[^3] by inserting sleeps at the appropriate points in the threading code. To do so, run the `repro_join_race.py` from the linked repo.

[^1]: https://github.com/python/cpython/blob/441affc9e7f419ef0b68f734505fa2f79fe653c7/Lib/threading.py#L1201
[^2]: https://github.com/python/cpython/blob/441affc9e7f419ef0b68f734505fa2f79fe653c7/Lib/threading.py#L1115
[^3]: https://github.com/mpage/cpython/commit/81946532792f938cd6f6ab4c4ff92a4edf61314f

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Linux

### PR DESCRIPTION
**Pull request by mpage**
This PR makes thread joining thread-safe in free-threaded builds and fixes gh-116372, which affects both default and free-threaded builds.

The combination of `_tstate_lock` and `_ThreadHandle` is replaced by `_ThreadHandle` and the supporting infrastructure for `_tstate_lock` is removed. Joining a thread now joins its `_ThreadHandle`.  The `_ThreadHandle`s for all threads that we want to join on shutdown (non-daemon threads that were created by the threading module) are tracked in a linked-list attached to the `_thread` module. At shutdown we iterate over this list and join the handles.

There is one wrinkle with this approach: the `_MainThread` instance. The `_MainThread` instance is joinable. Previously, it represented whatever thread imported the `threading` module. Typically this is the main thread of the main interpreter, but in some rare circumstances (embedding scenarios?) it might not be. The previous approach allowed the `MainThread` instance to be joined even if it wasn't the main thread of the main interpreter, but this isn't possible with the current approach. To work around this we change the `_MainThread` to always represent the main thread of the runtime, also fixing gh-83223.

* Issue: gh-114271

### CODE REVIEW

**Comment by mpage** [comment]
Just a heads up - please don't look at this until it's marked as ready for review. There are a couple of failing CI jobs that need to be addressed.

**Comment by mpage** [comment]
@colesbury - This is good to go. Would you please have a look?

**Comment by mpage** [comment]
Putting this back in draft. Going to see if I can move the join logic into `ThreadHandle`.

**Comment by mpage** [comment]
@colesbury - Could you take a look at this when you get a chance, please?

File: Lib/threading.py
Comment (colesbury) [inline Lib/threading.py:1141]: I think this needs to deal with the `self._handle == None` case. `_started` may be set before `self._handle` is set to a `ThreadHandle`.

File: Python/pystate.c
Comment (colesbury) [inline Python/pystate.c]: To fix the unused variable warning:

```suggestion
    assert(interp->threads.main == current_fast_get());
```

**Review by colesbury — COMMENTED** [review]
Overall, the approach looks nice!

I'm still looking this over, but I have some questions about `thread._handle`. Currently, `thread._handle` starts out as `None` and is then set to a `ThreadHandle` after the thread is started. This `None` -> `ThreadHandle` transition seems like it adds complexity and I think there may be a few bugs related to it (see comments below).

Would it be possible to always have a non-None `self._handle` by pushing a bit more state into `ThreadHandle`?

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: Probably worth mentioning that this also synchronizes calls to `set_done`

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c:36]: I think we should add a `PyMutex` to this struct to protect `shutdown_handles` rather than use `HEAD_LOCK(_PyRuntime);`

File: Lib/threading.py
Comment (mpage) [reply Lib/threading.py:1141]: Ah good catch!

**Comment by mpage** [comment]
> Would it be possible to always have a non-None self._handle by pushing a bit more state into ThreadHandle?

Yeah I think that's possible. By doing that I think we could also remove the `_started` event, too, but perhaps that should be a separate PR.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c:2197]: This looks like the loop used in the nogil forks, but I think we can do something simpler because the handles are in their own linked list:

Move the linked list to a local variable (clearing `state->shutdown_handles`) and then process the local linked list:

Roughly:

```c
HEAD_LOCK(&_PyRuntime);
struct llist_node list = LLIST_INIT(list);
list_concat(&list, &state->shutdown_handles);
HEAD_UNLOCK(&PyRuntime);

struct llist_node *node;
llist_for_each(node, &list) {
    ...
}
```

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: I think the language in the docs tends to be "if daemon is True" even when the parsing code coerces other values to `bool`.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: Should this be:

```
start_joinable_thread(function, daemon=True)\n\
```

Also, should the default be `daemon=True` to preserve existing behavior?

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: Should the default be `daemon = 1` to preserve existing behavior? I think that threads created by calling `start_joinable_thread()` directly did not previously delay shutdown.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: I think there's a possible race condition here between destruction of the `thread_module_state` during interpreter shutdown and this call. The `thread_module_state` containing the head of the linked list could already be deallocated. We can't fully rely on `thread_shutdown` because that can be interrupted by ctrl-c.

One possibility: clear the list in `thread_module_clear` so that `remove_from_shutdown_handles` is a no-op if the `thread_module_state` is cleared.

Comment (colesbury) [reply Modules/_threadmodule.c]: This might be a good reason to stick with `HEAD_LOCK(_PyRuntime);` instead of a mutex in `thread_module_state` for now.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: Consider naming this `PyThreadHandleObject` to further distinguish it from `ThreadHandle`.

File: Modules/_threadmodule.c
Comment (colesbury) [reply Modules/_threadmodule.c:36]: Or maybe not due to https://github.com/python/cpython/pull/114839#discussion_r1515128548

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c]: Yeah, this is the reason that I'm using `HEAD_LOCK(&_PyRuntime)` instead of a dedicated `PyMutex` but I missed the linked list case. Nice catch!

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:2197]: I don't think this ends up simplifying things. We'd still need to iterate over `list` while the lock is held, because the threads that are currently running may end up removing themselves.

**Review by colesbury — APPROVED** [review]
This looks good to me and I verified that it fixes a shutdown related crash when running with `test_threading.py` with `PYTHON_GIL=0`.

@gpshead - would you be interested in reviewing these changes as well?

**Comment by ericsnowcurrently** [comment]
> To work around this we change the _MainThread to always represent the main thread of the main interpreter.

For subinterpreters making it the main thread of the main interpreter doesn't seem right.  In my mental model in that case, it is the thread currently running main (i.e. `interp->threads.main`, via `_PyInterpreterState_SetRunningMain()`).  I haven't thought about the case where no thread is running main.  For the main interpreter it's a non-issue, of course.

File: Python/lock.c
Comment (ericsnowcurrently) [inline Python/lock.c:329]: These need to be removed from Include/internal/pycore_lock.h too, right?

File: Include/internal/pycore_lock.h
Comment (ericsnowcurrently) [inline Include/internal/pycore_lock.h]: This is already there a few lines up.

File: Lib/threading.py
Comment (ericsnowcurrently) [inline Lib/threading.py]: I'm assuming `_DummyThread` is sufficiently tested that we aren't risking breaking them somehow with this change.  The note above seems somewhat particular about their behavior, including the part about how we can' t wait for them.

**Comment by ericsnowcurrently** [comment]
Also see gh-83223.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Sorry I haven't reviewed this already.  It's a large, complex change that requires careful reading.  For future reference, splitting this up would have led to faster reviews.

I've gotten through everything except _threadmodule.c.  I'll work on that Monday, probably.

If I understood correctly, the bulk of this change (rather than the fundamental objective) is all about moving more implementation out of threading.py and into the thread handle in _threadmodule.c, especially what was the "tstate lock".  Is that right?

File: Lib/threading.py
Comment (mpage) [reply Lib/threading.py]: I'll flip this back, it just felt more natural to me.

File: Python/lock.c
Comment (mpage) [reply Python/lock.c:329]: Ugh, yes, sorry.

**Comment by mpage** [comment]
> For future reference, splitting this up would have led to faster reviews.

Yeah, sorry about the size. I'm not really sure this could have been split up.

> For subinterpreters making it the main thread of the main interpreter doesn't seem right.
> Also see https://github.com/python/cpython/issues/83223.

Sorry I'm probably not using the correct terminology. What's implemented in this PR is the behavior described in https://github.com/python/cpython/issues/83223.

> If I understood correctly, the bulk of this change (rather than the fundamental objective) is all about moving more implementation out of threading.py and into the thread handle in _threadmodule.c, especially what was the "tstate lock". Is that right?

Yep, that's correct. The functionality that `tstate_lock` provided is now in `_ThreadHandle` and waiting for threads to exit has moved to `thread_shutdown()`.

File: Lib/threading.py
Comment (ericsnowcurrently) [reply Lib/threading.py]: No worries.  I was just double-checking.

File: Modules/_threadmodule.c
Comment (ericsnowcurrently) [inline Modules/_threadmodule.c]: formatting:

```suggestion
static void thread_run(void *boot_raw);
struct bootstate;
static void thread_bootstate_free(struct bootstate *boot, int decref);
```

Also, it would be cleaner if we move the actual struct/function definitions up here.

Finally, let's move this to after the module state stuff rather than before.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c:1732]: I think we need to add the shutdown handle earlier, before we start the thread. At this point, it's possible that the thread has already exited and tried to remove itself the shutdown handle pool. If we add the shutdown handle here, we can end up triggering the assertion `assert(self->shutdown_node.next == NULL)` in `ThreadHandle_decref`.

This can get triggered by running `test_concurrent_futures_circular_import` with `PYTHON_GIL=0`.

File: Modules/_threadmodule.c
Comment (ericsnowcurrently) [inline Modules/_threadmodule.c:36]: What happens if this module gets reloaded?  Will all the outstanding handles get wiped out?

File: Modules/_threadmodule.c
Comment (ericsnowcurrently) [inline Modules/_threadmodule.c:1704]: I'm not sad to see this go. 😄

File: Modules/_threadmodule.c
Comment (ericsnowcurrently) [inline Modules/_threadmodule.c:2205]: FWIW, I find it surprising that the current (finalizing) thread might have a handle still.  That's probably something to solve separately though.

**Review by ericsnowcurrently — COMMENTED** [review]
Here are a few more comments.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:36]: They won't be present in the new module's list, so we wouldn't wait on them when shutdown is called. They would remain valid, however.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:2205]: This could happen if you called `os.fork()` in a thread created by the `threading` module and then called `_thread._shutdown()` in the child process.

File: Modules/_threadmodule.c
Comment (ericsnowcurrently) [reply Modules/_threadmodule.c:36]: That could be confusing, but I expect it is an *extremely* unlikely case.

**Comment by ericsnowcurrently** [comment]
FYI, I'm still having trouble feeling confident about any review I can give you on this.  I've spent some time poring over the threading module (and _threadmodule.c), in the interest of having a clearer picture.  In the process of that I ended up mostly doing some of the cleanup that I think would make this PR much easier to review.  You can see my "draft" changes on gh-116776.  

Ideally, this PR would be focused strictly on fixing the race, rather than including all the other changes you need in order to fix it, if that makes sense.

**Comment by mpage** [comment]
> Ideally, this PR would be focused strictly on fixing the race, rather than including all the other changes you need in order to fix it, if that makes sense.

@ericsnowcurrently - Yep, you'd like to see the necessary refactoring done as separate PRs. How would you like to proceed here? I think there is a smaller set of separable logical changes than what's in https://github.com/python/cpython/pull/116776 that might make this easier to review:

1. Fix `_MainThread`
2. Split `_ThreadHandle` into the Python wrapper and the separately refcounted part.
3. Consolidate the detached and joined states into a single done state.
4. Fix the race. This will still be a large PR but should be easier to review since it doesn't contain any of the necessary refactoring.

**Comment by ericsnowcurrently** [comment]
Yeah, you've captured it well.  FYI, I was looking at the `_MainThread` stuff separately already. 😄

**Comment by ericsnowcurrently** [comment]
I'd be glad to follow-through on the changes I posted in gh-116776, if that would help you.

**Comment by mpage** [comment]
> I'd be glad to follow-through on the changes I posted in https://github.com/python/cpython/pull/116776, if that would help you.

@ericsnowcurrently - Thanks! Is that something you think you can get done quickly (i.e. in the next few days)? If so, I'd be happy to stack on top of those changes once you're done.

**Comment by ericsnowcurrently** [comment]
Yeah, I'll aim to wrap it up tomorrow.

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c:1863]: Nit, but it's weird to have `daemon=True` as default, while we want to discourage daemon threads.
(admittedly this doesn't really matter since it's an internal API)

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c]: Nit: use infinitive
```suggestion
Wait for all non-daemon threads (other than the calling thread) to stop.");
```

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c]:
```suggestion
Internal only. Make a thread handle for a running thread not spawned\n
by the _thread or threading module.");
```

**Review by pitrou — APPROVED** [review]
+1. I really like the simplification here!

**Comment by ericsnowcurrently** [comment]
I'll defer to @pitrou's assessment on this.  I'd rather this land sooner than later and don't want my pace to hold things up.  Again, thanks for working on this.

(I'll follow up separately on any cleanup, about the "main" thread, and about our approach to managing thread lifetimes.  I actually appreciate not feeling so rushed to sort those out. 😄)

**Comment by ericsnowcurrently** [comment]
FWIW, this change feels a bit closer to what I had in mind last year: https://github.com/python/cpython/pull/110848#discussion_r1374989674.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:1863]: This was to keep it consistent with the previous behavior of the function.

**Comment by mpage** [comment]
The `CIFuzz (memory)` failure is unrelated to this diff.

**Comment by pitrou** [comment]
Do we have a Thread Sanitizer CI build somewhere? In my experience it's quite good at finding synchronization bugs.
@gpshead

**Comment by pitrou** [comment]
@gpshead Would you like to take a look at this?

**Comment by pitrou** [comment]
Ok, TSAN tests passed after merging from main (though there are too few of them currently).

**Comment by pitrou** [comment]
Updated again to get test_threading in the set of TSAN tests.

**Pull request merged**
